src/main/java/net/snowflake/client/core/SFSession.java

```diff
@@ -145,30 +144,51 @@ public QueryStatus getQueryStatus(String queryID) throws SQLException {
       statusUrl = sessionUrl + SF_PATH_QUERY_MONITOR + queryID;
     }
     // Create a new HTTP GET object and set appropriate headers
+
     HttpGet get = new HttpGet(statusUrl);
-    get.setHeader("Content-type", "application/json");
-    get.setHeader("Authorization", "Snowflake Token=\"" + this.sessionToken + "\"");
     String response = null;
     JsonNode jsonNode = null;
-    try {
-      response = HttpUtil.executeGeneralRequest(get, loginTimeout, getOCSPMode());
-      jsonNode = OBJECT_MAPPER.readTree(response);
-    } catch (Exception e) {
-      throw new SnowflakeSQLLoggedException(
-          this, e.getMessage(), "No response or invalid response from GET request. Error: {}");
-    }
-    // Get response as JSON and parse it to get the query status
-    // check the success field first
-    if (!jsonNode.path("success").asBoolean()) {
-      logger.debug("response = {}", response);
-
-      int errorCode = jsonNode.path("code").asInt();
-      throw new SnowflakeSQLException(
-          queryID,
-          jsonNode.path("message").asText(),
-          SqlState.SQLCLIENT_UNABLE_TO_ESTABLISH_SQLCONNECTION,
-          errorCode);
-    }
+    boolean sessionRenewed;
+    do {
```

Comment:
can you describe the logic in comments?

Reply:
In this loop, we attempt to execute an HTTP response to get back the query status. Previously, if the JSON response came back as a failure, we simply threw a SQLException detailing the failure.

Now, however, we check the error code of the failure first. 
If the error code corresponds to SESSION_EXPIRED_GS_CODE, we try to renew the session. 
While renewing the session, we check for a re-authentication error. If we get that, we try re-authenticating the session and then renewing. 
After renewing the session, we try to execute the original HTTP response again.
If we still come back with a failure, we finally throw a SQLException detailing this failure.

Reply:
Thanks for the explanation and comments! My next question is how do you test all those scenarios in the loop?

Reply:
These should be covered in the new class HeartbeatAsyncLatestIT.
First, the session token validity is reduced to 20 seconds and the master token validity is reduced to 60 seconds. There is a testSuccess() and testFailure() test. 

With testSuccess(), the heartbeat is enabled and we run an async query, sleep more than 60 seconds, and then attempt to retrieve the result.  In this case, initially the response will come back with an error code. The session will have to renew and try again. The second time, the response comes back with the query status successfully.

With testFailure, the heartbeat is not enabled and we run an async query, sleep more than 60 seconds, and then attempt to retrieve the result. The response comes back with an error code. We attempt to renew the session, but since heartbeat is not enabled, the session cannot be renewed and comes back with a re-authentication request failure. Since the authenticator is not an external browser, the error is returned as a SQLException. We parse this error and confirm it matches the reauthentication request exception.